docs/RFCS/monitoring_queries_and_sessions.md

```diff
@@ -0,0 +1,219 @@
+in the admin UI and sql shell, along with their start timestamps / durations. In this document,
+queries are defined as every executable statement; not just `SELECT`s. Everything that is converted
+into a `planNode` is a query.
+
+This RFC answers these use-cases:
+
+- Getting a list of node-local and cluster-wide active sessions through the CLI and the admin UI
+- Getting a list of node-local and cluster-wide executing queries through the CLI and the admin UI
+
+The following concerns are out of scope for this RFC:
+
+- Assigning unique identifiers to sessions and queries
+- Identifying and listing SQL transactions
+- Outlining a mechanism for cancelling sessions or queries 
+- The return values/errors returned by cancelled queries/sessions/closed connections.
+- Whether cancellation should be supported at SQL transaction level, or query/statement level.
+
+# Motivation
+
+Currently, there's no visibility into what queries/sessions are running on the cluster at any
+given point of time. Adding visibility into this has been a common customer request for
+a while.
+
+Albeit cancellation is out of scope for this RFC, being able to view sessions and queries
+along with information like application names, duration, etc, would be important information
+for DBAs to have, in order to make an appropriate cancellation decision (when that is implemented).
+
+Since some DB deployments can be shared between multiple client applications, some level of
+functionality to filter sessions and queries by client application names, both in the admin
+UI and in the CLI, would also be useful. And since not all DBAs are familiar with SQL,
+a non-SQL interface to monitor queries and sessions would be an asset.
+
+# Detailed design
+
+The inspiration behind some of the design comes from
+[this PR from Raphael](https://github.com/cockroachdb/cockroach/pull/10317) which implemented
+node-local session registries exposed via a virtual table.
+
+## Thread-safe session struct
+
+The session struct is currently not thread-safe; since all operations on it
+currently happen in one thread. Making it thread-safe involves making a nested struct
+inside it (called `mu`), similar to how the `Txn` struct is structured. There
+would be an `RWMutex` on this nested `mu` struct.
+
+These fields would be moved from the main `Session` struct, to the `mu` sub-struct:
+- User (name of user)
+- Remote address (IP, port)
+- Application Name
+- TxnState
+- Eventually: any contexts/object references useful for cancellation
+
+These two new fields will be added to `mu`:
+- Start timestamp
+- Active query (reference to Query object defined below, can be nil)
+
+The active query reference will contain a reference to a non-parallel query _if one is running_.
+If a set of parallelizable queries are running, they will all be in `parallelizeQueue`. Inspecting
+both the active query reference and the `parallelizeQueue` will be required to know all queries
+in flight.
```

Comment:
this is rather unfortunate. I wish we could model the parallelizeQueue as containing one query for non parallel query execution. Anyway this might not be something you fix.